File: Java/290.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] match=new String[26];
        int[] matchNum=new int[26];
        for(int i=0;i<26;i++)
            matchNum[i]=0;
        int len1=pattern.length(),len2=str.length();
        int k=0;
        for(int i=0;i<len1;i++)
        {           
            char ch=pattern.charAt(i);
            int j=(int)(ch-'a');
            StringBuffer s = new StringBuffer();
            while(k<len2&&str.charAt(k)!=' ')
            {
                s.append(str.charAt(k));
                k++;
            }
            k++;
            String word=s.toString();
            if(matchNum[j]==0)                
            {
                for(int y=0;y<26;y++)
                    if(matchNum[y]==1&&word.equals(match[y]))
                       return false;
                match[j]=word;
               matchNum[j]=1;
            }
            else
            {
                 if(!match[j].equals(word))
                     return false;
            }           
        }
        if(k-1==len2)
            return true;
        else
            return false;
    }
}
__________________________________________________________________________________________________
sample 32664 kb submission
class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] words = str.split(" ");
        // store char in pattern with matching word
        Map<Character, String> map = new HashMap<>();
        if (pattern.length() != words.length)
            return false;
        
        for (int i = 0; i < words.length; i++)
        {
            char curr = pattern.charAt(i);
            if (map.containsKey(curr))
            {
                // check if word for this char equals the current word
                if (!map.get(curr).equals(words[i]))
                    return false;
            }
            else
            {
                // if word already exists, then return false since it shouldn't be in map yet
                if (map.containsValue(words[i]))
                    return false;
                // put char with current word into map
                map.put(curr, words[i]);
            }
        }
        return true;
    }
}
__________________________________________________________________________________________________

File: Python3/290.py
__________________________________________________________________________________________________
sample 20 ms submission
class Solution:
    def wordPattern(self, pattern, string):
        store = {}
        done_list = []
        str_list = string.split()
        if len(str_list) != len(pattern) or string == "":
            return False
        for word_pair in zip(pattern,string.split()):
            if word_pair[0] in store:
                if store[word_pair[0]] != word_pair[1]:
                    return False
            else:
                if  word_pair[1] in done_list:
                    return False
                store[word_pair[0]] = word_pair[1]
                done_list.append(word_pair[1])
        return True
__________________________________________________________________________________________________
sample 13000 kb submission
class Solution:
    def wordPattern(self, pattern: str, str: str) -> bool:
        a=str.split()
        if len(pattern)!=len(a):
            return False
        dic={}
        
        for i in range(len(pattern)):
            if pattern[i] not in dic:
                if a[i] in dic.values():
                    return False
                dic[ pattern[i] ] = a[i]
            elif dic[ pattern[i] ] !=a[i]:
                return False
                
        return True
__________________________________________________________________________________________________
